### PR TITLE
Add REXReferral contract

### LINKED ISSUE #3: Referral System
This is an affiliate program. Fees paid by streamers will be shared with the affiliate that referred them.

This is used by people like FTX referrals links, something like: https://ricochet.exchange/?affiliateId=1

a9744597-9631-4629-b710-e7ec8f723d04

The 2% fee taken maybe further slip into portions for the affiliates that have referred volume to the app.

### PR DESCRIPTION
**Pull request by shreyaspapi**
Fixes #3

### CODE REVIEW

**Comment by nikramakrishnan** [comment]
We're left with the implementation for the distribution of referral rewards. @mikeghen you said this will be implemented in the contract extending the base (I couldn't find it in this repo, so I assume it's not written yet).

File: contracts/REXMarket.sol [hunk 1]
```diff
@@ -1,314 +1,372 @@
...
+
+    // Referral functions
+
+    function checkAffiliate(bytes memory _ctx) internal {
+        ISuperfluid.Context memory decompiledContext = host.decodeCtx(_ctx); 
+		// Decode userData
+        (string memory affiliateId) = abi.decode(decompiledContext.userData, (string));
+
+        safeRegisterUser(decompiledContext.msgSender, affiliateId);
+    }
```
Comment (shreyaspapi) [inline contracts/REXMarket.sol:172]: FYI, change is here.

File: contracts/REXMarket.sol [hunk 2]
```diff
@@ -1,314 +1,372 @@
...
+        onlyExpected(_superToken, _agreementClass)
+        returns (bytes memory newCtx)
+    {
+        newCtx = _ctx;
+
+        (address shareholder, int96 flowRate) = _getShareholderInfo(
+            _agreementData
+        );
+
+        checkAffiliate(newCtx);
```
Comment (shreyaspapi) [inline contracts/REXMarket.sol:196]: FYI, change is here.

File: contracts/REXReferral.sol [hunk 3]
```diff
@@ -23,6 +23,7 @@ contract REXReferral is Ownable {
     mapping(string => uint256) public affiliateIdToAffiliate;
     mapping(address => uint256) public userToAffiliate;
     mapping(address => bool) public isUserOrganic;
+    mapping(address=>bool) public allowedUsers;
```
Comment (mikeghen) [inline contracts/REXReferral.sol:26]: Can you not use "users" here and use `approvedApps`?

Comment (mikeghen) [inline contracts/REXReferral.sol:26]: Can you not use "users" here and use `approvedApps`?

File: contracts/REXReferral.sol [hunk 4]
```diff
@@ -55,6 +56,22 @@ contract REXReferral is Ownable {
...
+    // Whitelist an address
+    function whitelistAddress (address userAddr) public onlyOwner {
+        allowedUsers[userAddr]=true;
+    }
+
+    // Remove from whitelist
+    function removeFromWhitelistAddress (address userAddr) public onlyOwner {
+        allowedUsers[userAddr]=false;
+    } 
+
```
Comment (mikeghen) [inline contracts/REXReferral.sol:74]: Could you use Open zepplien `Roles.sol` like are used in REX Bank, like this PR shows: https://github.com/Ricochet-Exchange/rex-bank/pull/31

Roles has all the functionality implemented for adding/removing

There should be two roles: `Admin` and `ApprovedApp`

Also, whitelist should be "allowlist", but I think it you are using OZ Roles then you wont need to code these yourself

**Review by mikeghen — CHANGES_REQUESTED** [review]

File: contracts/REXReferral.sol (hunk 3 above)
Comment (shreyaspapi) [reply contracts/REXReferral.sol]: Done 👍

File: contracts/REXReferral.sol (hunk 4 above)
Comment (shreyaspapi) [reply contracts/REXReferral.sol]: Done 👍

File: contracts/REXReferral.sol (hunk 3 above)
Comment (nikramakrishnan) [reply contracts/REXReferral.sol]: This is not needed anymore 😄

**Review by mikeghen — CHANGES_REQUESTED** [review]

File: contracts/referral/REXReferral.sol [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+    modifier onlyApprovedApp() {
+        _checkRole(MARKET_CONTRACT_ROLE, msg.sender);
+        _;
+    }
+
+    // Add a market contract to allow-list
+    // OpenZeppelin already checks for admin role
+    function registerApp(address contractAddr) public {
+        grantRole(MARKET_CONTRACT_ROLE, contractAddr);
+    }
```
[the remark this answers is not in the document]
Comment (shreyaspapi) [reply contracts/referral/REXReferral.sol:79]: Inside `grantRole`, OpenZeppelin already checks for `admin`.

File: contracts/referral/REXReferral.sol [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+
+    // Remove a market contract from allow-list
+    // OpenZeppelin already checks for admin role
+    function unregisterApp(address contractAddr) public {
+        revokeRole(MARKET_CONTRACT_ROLE, contractAddr);
+    }
```
[the remark this answers is not in the document]
Comment (shreyaspapi) [reply contracts/referral/REXReferral.sol:85]: Inside grantRole, OpenZeppelin already checks for admin, do we still want that here?

File: contracts/referral/REXReferral.sol [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+        public
+        validAffiliateAddress
+        notZero(newAddress)
+    {
+        uint256 affiliateIdx = addressToAffiliate[msg.sender];
+        Affiliate storage affiliate = affiliates[affiliateIdx];
+        affiliate.addr = newAddress;
+        delete addressToAffiliate[msg.sender];
+        addressToAffiliate[newAddress] = affiliateIdx;
+    }
```
[the remark this answers is not in the document]
Comment (nikramakrishnan) [reply contracts/referral/REXReferral.sol:149]: Won't the contract that distributes the exchanged token have to call `getAffiliateAddress` in REXReferral to get the affiliate's address? Having this here will allow a referrer to change the address where they want to receive rewards (this is assuming we're still using the 'distribute every 30 minutes' logic).

[the remark this answers is not in the document]
Comment (shreyaspapi) [reply contracts/referral/REXReferral.sol:149]: resolved by convo in discord.

File: contracts/referral/REXReferral.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (nikramakrishnan) [reply contracts/referral/REXReferral.sol:79]: Added onlyAdmin here for clarity.

File: contracts/referral/REXReferral.sol (hunk 6 above)
[the remark this answers is not in the document]
Comment (nikramakrishnan) [reply contracts/referral/REXReferral.sol:84]: Added onlyAdmin here for clarity.

**Comment by shreyaspapi** [comment]
Tests changed by @harshitsinghai77, ready for review 🚀

**Review by mikeghen — APPROVED** [review]

**Pull request merged**